sor/src/main/java/com/bazaarvoice/emodb/sor/db/astyanax/CqlDataWriterDAO.java

```diff
@@ -0,0 +1,198 @@
+            batchStatement.add(QueryBuilder.insertInto(tableDDL.getTableMetadata())
+                    .value(tableDDL.getRowKeyColumnName(), rowKey)
+                    .value(tableDDL.getChangeIdColumnName(), changeId)
+                    .value(tableDDL.getBlockColumnName(), i)
+                    .value(tableDDL.getValueColumnName(), blocks.get(i))
+                    .setConsistencyLevel(consistencyLevel));
+        }
+    }
+
+    private void writeCompaction(ByteBuffer rowKey, UUID compactionKey, Compaction compaction,
+                                 WriteConsistency consistency, DeltaPlacement placement,
+                                 CassandraKeyspace keyspace, Table table, String key) {
+
+        // Add the compaction record
+        ByteBuffer encodedCompaction = ByteBuffer.wrap(_changeEncoder.encodeCompaction(compaction, new StringBuilder(_deltaPrefix)).toString().getBytes(Charsets.UTF_8));
+
+        Session session = keyspace.getCqlSession();
+        ConsistencyLevel consistencyLevel = SorConsistencies.toCql(consistency);
+        DeltaTableDDL tableDDL = placement.getDeltaTableDDL();
+
+        BatchStatement batchStatement = new BatchStatement(BatchStatement.Type.LOGGED);
+
+        insertBlockedDeltas(batchStatement, tableDDL, consistencyLevel, rowKey, compactionKey, encodedCompaction);
+
+        session.execute(batchStatement);
+    }
+
+    private void deleteCompactedDeltas(ByteBuffer rowKey, WriteConsistency consistency, DeltaPlacement placement,
+                                       CassandraKeyspace keyspace, Collection<UUID> changesToDelete,
+                                       List<History> historyList, Table table, String key) {
+
+        Session session = keyspace.getCqlSession();
+        ConsistencyLevel consistencyLevel = SorConsistencies.toCql(consistency);
+
+        // delete the old deltas & compaction records
+
+        DeltaTableDDL tableDDL = placement.getDeltaTableDDL();
+
+        // each individual delete statement will be atomic, so the batch can be unlogged
+        // if any delete fails in this batch it will be deleted again the next time it is read
+        BatchStatement batchStatement = new BatchStatement(BatchStatement.Type.UNLOGGED);
+
+        for (UUID change : changesToDelete) {
+            batchStatement.add(QueryBuilder.delete()
+                    .from(tableDDL.getTableMetadata())
+                    .where(eq(tableDDL.getRowKeyColumnName(), rowKey))
+                    .and(eq(tableDDL.getChangeIdColumnName(), change))
+                .setConsistencyLevel(consistencyLevel));
+        }
+
+
+        if (historyList != null && !historyList.isEmpty()) {
+            AuditBatchPersister auditBatchPersister = CqlAuditBatchPersister.build(batchStatement, placement.getDeltaHistoryTableDDL(),
+                    _changeEncoder, _auditStore);
+            _auditStore.putDeltaAudits(rowKey, historyList, auditBatchPersister);
```

Comment:
Doesn't this violate your findings about using CQL batch statements? This approach results in writes to two tables in the same batch statement: "xxx_delta" and "xxx_history". Generally audits are a best-effort write, as seen [here](https://github.com/bazaarvoice/emodb/blob/master/sor/src/main/java/com/bazaarvoice/emodb/sor/core/DefaultAuditStore.java#L45). Maybe it makes sense for the CQL implementation to just do a blind async write with consistency level ANY instead of combining with the batch statement. Thoughts?

Reply:
done in migration

Reply:
done